src/util.rs

```diff
@@ -16,13 +18,15 @@ pub(crate) fn witness_size(wit: &[Vec<u8>]) -> usize {
     wit.iter().map(Vec::len).sum::<usize>() + varint_len(wit.len())
 }
 
-pub(crate) fn witness_to_scriptsig(witness: &[Vec<u8>]) -> Script {
+pub(crate) fn witness_to_scriptsig(witness: &[Vec<u8>]) -> ScriptBuf {
     let mut b = script::Builder::new();
     for wit in witness {
         if let Ok(n) = script::read_scriptint(wit) {
             b = b.push_int(n);
         } else {
-            b = b.push_slice(wit);
+            // FIXME: There has to be a better way than this.
+            let push = PushBytesBuf::try_from(wit.clone()).expect("FIXME: Handle error");
+            b = b.push_slice(&push)
```

Comment:
Hey @Kixunil can you take a look at this please. I cannot work out how to get something that implements `AsRef<PushBytes>` from a `&Vec<u8>` to pass to `push_slice`? Only using a reference and checking that the vec is a valid length before passing it to `push_slice`  is exactly what `PushBytes` is for right?

Reply:
Firstly, I don't understand why this function even exists - witnesses and scripts are different.

The only way is to return `Option` or `Result` from the function or document that it panics if a `witness` with too large element is passed in.

Also to avoid allocation:

```rust
let push: &PushBytes = wit.try_into()?; // or expect
b = b.push_slice(&push);
```

Reply:
This is an internal to how rust-miniscript works. We have a function that constructs a `Vec<Vec<u8>>` type for all types of scripts. If the script is legacy p2sh or bare script pubkey, we need to convert this to `ScriptBuf` converting each `Vec<u8>` into a scriptPush. 

> The only way is to return Option or Result from the function or document that it panics if a witness with too large element is passed in.

This is one of the selling points of miniscript. Miniscripts are statically analyzed to make sure that all valid spend paths constructed at spend time are within bounds and spendable on bitcoin network. In other words, this function is only called on satisfactions generated from miniscript satisfier which guarantees the resource bounds enforced by bitcoin consensus rules.

It does context based checks that total witness length cannot be more than 3600 for segwitv0 and so on. 

> What KixUnil is saying is correct. 

The allocation should be avoided. I don't remember how I did this :( .

Reply:
Would it make sense to pass in something like `&[WitnessElement]` where `WitnessElement` is a type that guarantees length at most 3600? You could then safely implement `AsRef<PushBytes>` for it. Or if it's not that important for miniscript to carry the precise bound just use `&[PushBytesBuf]`.

Reply:
> Firstly, I don't understand why this function even exists - witnesses and scripts are different.

Prior to segwit they are not.

Reply:
Prior to segwit witness doesn't exist.

Reply:
We could use `&PushBytesBuf`, since all witnesses we push are either signatures, hash preimages, or constants like 1 or 0.